File: Medium/82.删除排序链表中的重复元素-ii.cpp
/*
 * @lc app=leetcode.cn id=82 lang=cpp
 *
 * [82] 删除排序链表中的重复元素 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *deleteDuplicates(ListNode *head)
    {
        //特殊情况
        if (!head || !head->next)
            return head;
        //新建一个哑节点用于统一运算（不然就要特殊考虑第一个节点就应该删除的情况）
        ListNode *dummyhead = new ListNode(INT_MAX);
        dummyhead->next = head;
        ListNode *pre = dummyhead;
        ListNode *cur = head;
        while (cur && cur->next)
        {
            //初始化的时pre指向的是哑结点，所以比较逻辑应该是pre的下一个节点和cur的下一个节点
            if (pre->next->val != cur->next->val)
            {
                pre = pre->next;
                cur = cur->next;
            }
            else
            {
                //如果pre、cur指向的节点值相等，就不断移动cur，直到pre、cur指向的值不相等
                while (cur && cur->next && pre->next->val == cur->next->val)
                    cur = cur->next;
                // pre->next跳过中间所有的重复元素
                pre->next = cur->next;
                cur = cur->next;
            }
        }
        return dummyhead->next;
    }
};
// @lc code=end

File: Medium/82.删除排序链表中的重复元素-ii.go
/*
 * @lc app=leetcode.cn id=82 lang=golang
 *
 * [82] 删除排序链表中的重复元素 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
	// 虚节点来统一操作
	dummyNode := &ListNode{Val: -1}
	dummyNode.Next = head
	// cur指向前一个没有重复的节点
	cur := dummyNode
	// 循环
	for cur.Next != nil && cur.Next.Next != nil {
		// 如果后续出现重复节点
		if cur.Next.Val == cur.Next.Next.Val {
			// 循环删除每个重复的节点
			temp := cur.Next.Val
			for cur.Next != nil && cur.Next.Val == temp {
				cur.Next = cur.Next.Next
			}
			// 更新前一个没有重复的节点
		} else {
			cur = cur.Next
		}
	}
	return dummyNode.Next
}

// @lc code=end

